tests/activity_api_pause_test.go

```diff
@@ -0,0 +1,461 @@
+	s.NoError(err)
+	s.Equal(1, len(description.PendingActivities))
+	s.True(description.PendingActivities[0].Paused)
+	s.Equal(int32(1), description.PendingActivities[0].Attempt)
+
+	// unpause the activity
+	unpauseRequest := &workflowservicepb.UnpauseActivityByIdRequest{
+		Namespace:  s.Namespace(),
+		WorkflowId: workflowRun.GetID(),
+		ActivityId: "activity-id",
+		Operation: &workflowservicepb.UnpauseActivityByIdRequest_Resume{
+			Resume: &workflowservicepb.UnpauseActivityByIdRequest_ResumeOperation{
+				NoWait: false,
+			},
+		},
+	}
+	unpauseResp, err := s.FrontendClient().UnpauseActivityById(ctx, unpauseRequest)
+	s.NoError(err)
+	s.NotNil(unpauseResp)
+
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		assert.Equal(t, int32(2), completedActivityCount.Load())
+	}, 5*time.Second, 100*time.Millisecond)
+
+	var out string
+	err = workflowRun.Get(ctx, &out)
+
+	s.NoError(err)
+}
+
+func (s *ActivityApiPauseClientTestSuite) TestActivityPauseApi_WhileWaiting() {
+	// In this case, pause happens when activity is in retry state.
+	// Make sure that activity is paused and then unpaused.
+	// Also check that activity will not be retried while unpaused.
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	s.initialRetryInterval = 1 * time.Second
+	s.activityRetryPolicy = &temporal.RetryPolicy{
+		InitialInterval:    s.initialRetryInterval,
+		BackoffCoefficient: 1,
+	}
+
+	var completedActivityCount atomic.Int32
+
+	activityFunction := func() (string, error) {
+		completedActivityCount.Add(1)
+		if completedActivityCount.Load() == 1 {
+			activityErr := errors.New("bad-luck-please-retry")
+			return "", activityErr
+		}
+		return "done!", nil
+	}
+
+	workflowFn := s.makeWorkflowFunc(activityFunction)
+
+	s.Worker().RegisterWorkflow(workflowFn)
+	s.Worker().RegisterActivity(activityFunction)
+
+	workflowOptions := sdkclient.StartWorkflowOptions{
+		ID:        testcore.RandomizeStr("wf_id-" + s.T().Name()),
+		TaskQueue: s.TaskQueue(),
+	}
+
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	// wait for activity to start
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, int32(1), completedActivityCount.Load())
+	}, 5*time.Second, 500*time.Millisecond)
+
+	// pause activity
+	pauseRequest := &workflowservicepb.PauseActivityByIdRequest{
+		Namespace:  s.Namespace(),
+		WorkflowId: workflowRun.GetID(),
+		ActivityId: "activity-id",
+	}
+	resp, err := s.FrontendClient().PauseActivityById(ctx, pauseRequest)
+	s.NoError(err)
+	s.NotNil(resp)
+
+	// wait long enough for activity to retry if pause is not working
+	util.InterruptibleSleep(ctx, 2*time.Second)
```

Comment:
When you get here, you may already be in the second attempt. You have no guarantees on timing.

You can assume that after 2 seconds the activity is probably no longer in the process of being rescheduled though and check as a baseline that after another 2 second sleep, no new attempt is scheduled.

Reply:
for this to happen it should be thread switch right before PauseActivityByIdRequest, and don't come back for more then a second. Yes, in this case it will fail. But as far as I know we don't have a better way. 
To be sure that 2nd retry will not happen we need "test mixins" - code that will block 2nd retry in pause is not called yet.
Any proposal?